testutils/testtracing/propagation.go

```diff
@@ -0,0 +1,313 @@
+type PropagationTestCase struct {
+	ForwardCount      int
+	TracingDisabled   bool
+	ExpectedBaggage   string
+	ExpectedSpanCount int
+}
+
+type tracerChoice struct {
+	tracerType       TracerType
+	tracer           opentracing.Tracer
+	spansRecorded    func() int
+	resetSpans       func()
+	isFake           bool
+	zipkinCompatible bool
+}
+
+// Run executes the test cases in the test suite against 3 different tracer implementations
+func (s *PropagationTestSuite) Run(t *testing.T) {
+	tests := []struct {
+		name string
+		run  func(t *testing.T)
+	}{
+		{"Noop_Tracer", s.runWithNoopTracer},
+		{"Mock_Tracer", s.runWithMockTracer},
+		{"Jaeger_Tracer", s.runWithJaegerTracer},
+	}
+	for _, test := range tests {
+		t.Logf("Running with %s", test.name)
+		test.run(t)
+	}
+}
+
+func (s *PropagationTestSuite) runWithNoopTracer(t *testing.T) {
+	s.runWithTracer(t, tracerChoice{
+		tracer:        nil, // will cause opentracing.GlobalTracer() to be used
+		tracerType:    Noop,
+		spansRecorded: func() int { return 0 },
+		resetSpans:    func() {},
+		isFake:        true,
+	})
+}
+
+func (s *PropagationTestSuite) runWithMockTracer(t *testing.T) {
+	mockTracer := mocktracer.New()
+	s.runWithTracer(t, tracerChoice{
+		tracerType: Mock,
+		tracer:     mockTracer,
+		spansRecorded: func() int {
+			return len(MockTracerSampledSpans(mockTracer))
+		},
+		resetSpans: func() {
+			mockTracer.Reset()
+		},
+	})
+}
+
+func (s *PropagationTestSuite) runWithJaegerTracer(t *testing.T) {
+	jaegerReporter := jaeger.NewInMemoryReporter()
+	jaegerTracer, jaegerCloser := jaeger.NewTracer(testutils.DefaultServerName,
+		jaeger.NewConstSampler(true),
+		jaegerReporter)
+	// To enable logging, use composite reporter:
+	// jaeger.NewCompositeReporter(jaegerReporter, jaeger.NewLoggingReporter(jaeger.StdLogger)))
+	defer jaegerCloser.Close()
+	s.runWithTracer(t, tracerChoice{
+		tracerType: Jaeger,
+		tracer:     jaegerTracer,
+		spansRecorded: func() int {
+			return len(jaegerReporter.GetSpans())
+		},
+		resetSpans: func() {
+			jaegerReporter.Reset()
+		},
+		zipkinCompatible: true,
+	})
+}
+
+func (s *PropagationTestSuite) runWithTracer(t *testing.T, tracer tracerChoice) {
+	testCases, ok := s.TestCases[tracer.tracerType]
+	if !ok {
+		t.Logf("No test cases for encoding=%s and tracer=%s", s.Encoding.Format, tracer.tracerType)
+		return
+	}
+	opts := &testutils.ChannelOpts{
+		ChannelOptions: tchannel.ChannelOptions{Tracer: tracer.tracer},
+		DisableRelay:   true,
+	}
+	ch := testutils.NewServer(t, opts)
+	defer ch.Close()
+	ch.Peers().Add(ch.PeerInfo().HostPort)
+	call := s.Register(t, ch)
+	for _, tt := range testCases {
+		s.runTestCase(t, tracer, ch, tt, call)
+	}
+}
+
+func (s *PropagationTestSuite) runTestCase(
+	t *testing.T,
+	tracer tracerChoice,
+	ch *tchannel.Channel,
+	test PropagationTestCase,
+	call TracingCall,
+) {
+	descr := fmt.Sprintf("test %+v with tracer %+v", test, tracer)
+	ch.Logger().Debug("")
+	ch.Logger().Debugf("======> STARTING %s", descr)
```

Comment:
do we need these logs now that the test development is complete and it works? Maybe just a simple "Starting tracing test: %v"? I'd like to avoid the ASCII art though